core/src/queue.rs

```diff
@@ -249,21 +279,115 @@ impl Queue {
             return;
         }
 
-        let mut seen = Vec::new();
+        let mut seen_queue = Vec::new();
+        let mut seen_waiting_buffer = Vec::new();
+
+        let txs_from_queue = core::iter::from_fn(|| self.pop_from_queue(&mut seen_queue, wsv));
+        let txs_from_waiting_buffer =
+            core::iter::from_fn(|| self.pop_from_waiting_buffer(&mut seen_waiting_buffer, wsv));
 
         let transactions_hashes: HashSet<HashOf<VersionedSignedTransaction>> = transactions
             .iter()
             .map(VersionedAcceptedTransaction::hash)
             .collect();
-        let out = std::iter::from_fn(|| self.pop(&mut seen, wsv))
+        let txs = txs_from_queue
+            .round_robin(txs_from_waiting_buffer)
             .filter(|tx| !transactions_hashes.contains(&tx.hash()))
             .take(self.txs_in_block - transactions.len());
-        transactions.extend(out);
+        transactions.extend(txs);
+
+        for (seen, queue) in [
+            (seen_queue, &self.queue),
```

Comment:
This function is an unholy mix of functional and imperative. Think we can rewrite parts of it to be one or the other?

Reply:
Hope it looks better now